demos/common/hfsm/az_hfsm.c

```diff
@@ -0,0 +1,197 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file az_hfsm.c
+ * @brief Hierarchical Finite State Machine implementation.
+ */
+
+#include <stdint.h>
+#include <stddef.h>
+#include <azure/core/internal/az_precondition_internal.h>
+
+#include "az_hfsm.h"
+
+const az_hfsm_event az_hfsm_entry_event = { AZ_HFSM_ENTRY, NULL };
+const az_hfsm_event az_hfsm_exit_event = { AZ_HFSM_EXIT, NULL };
+const az_hfsm_event az_hfsm_timeout_event = { AZ_HFSM_TIMEOUT, NULL };
+const az_hfsm_event az_hfsm_errork_unknown_event = { AZ_HFSM_ERROR, NULL };
+
+/**
+ * @brief Initializes the HFSM.
+ * 
+ * @param[in] h The HFSM handle.
+ * @param root_state The root state for this HFSM.
+ * @param get_parent_func The function describing the HFSM structure.
+ * @return int32_t 
+ */
+int32_t az_hfsm_init(az_hfsm* h, az_hfsm_state_handler root_state, az_hfsm_get_parent get_parent_func)
+{
+  _az_PRECONDITION_NOT_NULL(h);
+  h->current_state = root_state;
+  h->get_parent_func = get_parent_func;
```

Comment:
looks like get_parent_func can't be null, add to precondition?

Reply:
Correct - both root_state and get_parent_func added to preconditions.